test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller_phases.go

```diff
@@ -0,0 +1,302 @@
+}
+
+// getDockerMachinesToDelete returns a list of DockerMachines representing excess replicas and Docker containers that do not match the infrastructure spec.
+// In the case of excess replicas, DockerMachines with the clusterv1.DeleteMachineAnnotation are prioritized to ensure they are deleted first.
+func getDockerMachinesToDelete(ctx context.Context, dockerMachines []infrav1.DockerMachine, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool) ([]infrav1.DockerMachine, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	dockerMachinesToDelete := []infrav1.DockerMachine{}
+	labelFilters := map[string]string{dockerMachinePoolLabel: dockerMachinePool.Name}
+
+	// Sort priority delete to end of list
+	sort.Slice(dockerMachines, func(i, j int) bool {
+		_, iHasAnnotation := dockerMachines[i].Annotations[clusterv1.DeleteMachineAnnotation]
+		_, jHasAnnotation := dockerMachines[j].Annotations[clusterv1.DeleteMachineAnnotation]
+
+		if iHasAnnotation && jHasAnnotation {
+			return dockerMachines[i].Name < dockerMachines[j].Name
+		}
+
+		return jHasAnnotation
+	})
+
+	desiredReplicas := int(*machinePool.Spec.Replicas)
+	totalNumMachines := 0
+	for _, dockerMachine := range dockerMachines {
+		totalNumMachines++
+		if totalNumMachines > desiredReplicas {
+			dockerMachinesToDelete = append(dockerMachinesToDelete, dockerMachine)
+			log.V(2).Info("Selecting DockerMachine for deletion because it is an excess replica", "dockerMachine", dockerMachine.Name, "namespace", dockerMachine.Namespace)
+			totalNumMachines--
+		} else {
+			externalMachine, err := docker.NewMachine(ctx, cluster, dockerMachine.Name, labelFilters)
+			if err != nil {
+				return nil, err
+			}
+			if !isMachineMatchingInfrastructureSpec(ctx, externalMachine, machinePool, dockerMachinePool) {
+				log.V(2).Info("Selecting DockerMachine for deletion because it does not match infrastructure spec", "dockerMachine", dockerMachine.Name)
+				dockerMachinesToDelete = append(dockerMachinesToDelete, dockerMachine)
+				totalNumMachines--
+			} else {
+				log.V(2).Info("Keeping DockerMachine, nothing to do", "dockerMachine", dockerMachine.Name, "namespace", dockerMachine.Namespace)
+			}
+		}
+	}
+
+	return dockerMachinesToDelete, nil
+}
+
+// isMachineMatchingInfrastructureSpec returns true if the Docker container image matches the custom image in the DockerMachinePool spec.
+func isMachineMatchingInfrastructureSpec(ctx context.Context, machine *docker.Machine, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool) bool {
+	_ = ctrl.LoggerFrom(ctx)
+	// NOTE: With the current implementation we are checking if the machine is using a kindest/node image for the expected version,
+	// but not checking if the machine has the expected extra.mounts or pre.loaded images.
+
+	semVer, err := semver.Parse(strings.TrimPrefix(*machinePool.Spec.Template.Spec.Version, "v"))
+	if err != nil {
+		// TODO: consider if to return an error
```

Comment:
This TODO was already there, I just copied this function over from nodepool.go after refactoring.